drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,815 @@
+ */
+static void i2c_omap_transmit_receive_data(const struct device *dev, uint8_t num_bytes)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint8_t *buf_ptr = data->current_msg.buf;
+
+	while (num_bytes--) {
+		if (data->receiver) {
+			*buf_ptr++ = i2c_base_addr->DATA;
+		} else {
+			i2c_base_addr->DATA = *(buf_ptr++);
+		}
+	}
+}
+
+/**
+ * @brief Resize the FIFO buffer for the OMAP I2C controller.
+ *
+ * This function resizes the FIFO buffer for the OMAP I2C controller based on the specified size.
+ * It clears the RX threshold and sets the new size for the receiver, or clears the TX threshold
+ * and sets the new size for the transmitter.
+ *
+ * @param dev Pointer to the device structure.
+ * @param size The new size of the FIFO buffer.
+ */
+static void i2c_omap_resize_fifo(const struct device *dev, uint8_t size)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint16_t buf_cfg;
+
+	buf_cfg = i2c_base_addr->BUF;
+	if (data->receiver) {
+		buf_cfg |= ((size) << 8) | I2C_OMAP_BUF_RXFIF_CLR;
+	} else {
+		buf_cfg |= (size) | I2C_OMAP_BUF_TXFIF_CLR;
+	}
+	i2c_base_addr->BUF = buf_cfg;
+}
+
+#ifdef CONFIG_I2C_OMAP_BUS_RECOVERY
+/**
+ * @brief Get the state of the SDA line.
+ *
+ * This function retrieves the state of the SDA (data) line for the OMAP I2C controller.
+ *
+ * @param io_context The I2C context.
+ * @return The state of the SDA line.
+ */
+static int i2c_omap_get_sda(void *io_context)
+{
+	const struct i2c_omap_cfg *cfg = (const struct i2c_omap_cfg *)io_context;
+	i2c_omap_regs_t *i2c_base_addr = (i2c_omap_regs_t *)cfg->base.addr;
+	uint32_t reg = i2c_base_addr->SYSTEST;
+
+	return reg & I2C_OMAP_SYSTEST_SDA_I_FUNC;
+}
+
+/**
+ * @brief Set the state of the SDA line.
+ *
+ * This function sets the state of the SDA (data) line for the OMAP I2C controller.
+ *
+ * @param io_context The I2C context.
+ * @param state The state to set (0 for low, 1 for high).
+ */
+static void i2c_omap_set_sda(void *io_context, int state)
+{
+	const struct i2c_omap_cfg *cfg = (const struct i2c_omap_cfg *)io_context;
+	i2c_omap_regs_t *i2c_base_addr = (i2c_omap_regs_t *)cfg->base.addr;
+	uint32_t reg = i2c_base_addr->SYSTEST;
+
+	if (state) {
+		reg |= I2C_OMAP_SYSTEST_SDA_O;
+	} else {
+		reg &= ~I2C_OMAP_SYSTEST_SDA_O;
+	}
+	i2c_base_addr->SYSTEST = reg;
+}
+
+/**
+ * @brief Set the state of the SCL line.
+ *
+ * This function sets the state of the SCL (clock) line for the OMAP I2C controller.
+ *
+ * @param io_context The I2C context.
+ * @param state The state to set (0 for low, 1 for high).
+ */
+static void i2c_omap_set_scl(void *io_context, int state)
+{
+	const struct i2c_omap_cfg *cfg = (const struct i2c_omap_cfg *)io_context;
+	i2c_omap_regs_t *i2c_base_addr = (i2c_omap_regs_t *)cfg->base.addr;
+	uint32_t reg = i2c_base_addr->SYSTEST;
+
+	if (state) {
+		reg |= I2C_OMAP_SYSTEST_SCL_O;
```

Comment:
You might be able to just avoid the temp variable `reg`, directly set/unset the bit, for instance here:

`i2c_base_addr->SYSTEST |= I2C_OMAP_SYSTEST_SCL_O;`

Same above in `i2c_omap_set_sda()` and `i2c_omap_resize_fifo()`